gnovm/pkg/test/filetest.go

```diff
@@ -380,65 +412,108 @@ func (d Directives) FileTest() string {
 	for sc.Scan() {
-		// Re-append trailing newline.
-		// Useful as we always use it anyway.
-		txt := sc.Text() + "\n"
+		last := &parsed[len(parsed)-1]
+		txt := sc.Text()
 		if !strings.HasPrefix(txt, "//") {
-			if len(parsed) == 0 || parsed[len(parsed)-1].Name != "" {
-				parsed = append(parsed, Directive{Content: txt})
+			// If we're already in an incomplete text directive, simply append there.
+			if last.Name == "" && !last.Complete {
+				last.Content += txt + "\n"
+				last.LastLine = txt
 				continue
 			}
-			parsed[len(parsed)-1].Content += txt
+			// Otherwise make a new directive.
+			parsed = append(parsed,
+				Directive{
+					Content:  txt + "\n",
+					LastLine: txt,
+				})
 			continue
 		}
 
-		comment := txt[2 : len(txt)-1]             // leading double slash, trailing \n
+		comment := txt[2:]                         // leading double slash
 		comment = strings.TrimPrefix(comment, " ") // leading space (if any)
 
-		// If we're already in a directive, simply append there.
-		if len(parsed) > 0 && parsed[len(parsed)-1].Name != "" {
-			parsed[len(parsed)-1].Content += comment + "\n"
-			continue
+		// Special case if following an incomplete comment line,
+		// always append to it even if it looks like `// TODO: ...`.
+		if strings.HasPrefix(txt, "//") &&
+			strings.HasPrefix(last.LastLine, "//") &&
+			!last.Complete {
+
+			if last.Name == "" {
+				// Just append text to it.
+				last.Content += txt + "\n"
+				last.LastLine = txt
+				continue
+			} else {
+				// Just append comment to it.
+				last.Content += comment + "\n"
+				last.LastLine = txt
+				continue
+			}
 		}
 
 		// Find if there is a colon (indicating a possible directive).
 		subm := reDirectiveLine.FindStringSubmatch(comment)
 		switch {
-		case subm == nil:
-			// Not found; append to parsed as a line, or to the previous
-			// directive if it exists.
-			if len(parsed) == 0 {
-				parsed = append(parsed, Directive{Content: txt})
-				continue
-			}
-			last := &parsed[len(parsed)-1]
-			if last.Name == "" {
-				last.Content += txt
-			} else {
-				last.Content += comment + "\n"
+		case subm == nil: // comment...
+			// If we're already in an incomplete directive, simply append there.
+			if !last.Complete {
+				if last.Name == "" {
+					last.Content += txt + "\n"
+					last.LastLine = txt
+					continue
+				} else {
+					last.Content += comment + "\n"
+					last.LastLine = txt
+					continue
+				}
 			}
-		case subm[1] != "": // output directive, with content on newlines
-			parsed = append(parsed, Directive{Name: subm[1]})
-		default: // subm[2] != "", all caps
+			// Otherwise make a new directive.
+			parsed = append(parsed,
+				Directive{
+					Content:  txt + "\n",
+					LastLine: txt,
+				})
+		case subm[1] != "": // CamelCase:
+			// output directive, with content on newlines
 			parsed = append(parsed,
-				Directive{Name: subm[2], Content: subm[3]},
-				// enforce new directive later
-				Directive{},
-			)
```

Comment:
This was producing too many directives. 
The new implementation makes appropriate groupings. 
It's both kinda confusing but I think the right is slightly easier to reason about for me.